docs/design/paginators.md

```diff
@@ -68,71 +68,87 @@ structure FunctionConfiguration { ... }
+fun <T> SdkAsyncIterable<T?>.filterNotNull(): SdkAsyncIterable<T> = TODO("not-implemented")
+
+suspend fun<T> SdkAsyncIterable<T>.toList(): List<T> = TODO("not-implemented")
+suspend fun<T> SdkAsyncIterable<T>.toSet(): Set<T> = TODO("not-implemented")
+
+/**
+ * Terminal operator that collects the given iterable and calls [action] for each result
+ */
+inline fun <T> SdkAsyncIterable<T>.collect(crossinline action: suspend (T) -> Unit):Unit = TODO()
 ```
 
-Pagination by default will always generate a paginator _over the normal operation output type_. If `items` is specified then the generated paginator will be “specialized” to have an additional field that paginates only over the member targeted by `items`. An example of this is demonstrated below:
+Pagination by default will always generate an `SdkAsyncIterable` over the normal operation output type. 
+If `items` is specified then the generated paginator will be “specialized” to have an additional field that iterates 
+only over the member targeted by `items`. The targeted type will be flattened for the user automatically.
+
+An example of this is demonstrated below:
 
 
 ```kotlin
 class ListFunctionsPaginator(
     private val client: LambdaClient,
     private val initialRequest: ListFunctionsRequest
-): Paginator<ListFunctionsResponse>{
-    private var cursor: String? = null
-    private var isFirstPage: Boolean = true
+): SdkAsyncIterable<ListFunctionsResponse>, SdkAsyncIterator<ListFunctionsResponse> {
+   private var cursor: String? = null
+   private var isFirstPage: Boolean = true
 
-    override val hasMorePages: Boolean
-        get() = isFirstPage || (cursor?.isNotEmpty() ?: false)
+   override operator fun iterator(): SdkAsyncIterator<ListFunctionsResponse> = this
 
-    override suspend fun next(): ListFunctionsResponse {
-        if (!hasMorePages) throw NoSuchElementException("no pages remaining")
+   override operator fun hasNext(): Boolean
+           = isFirstPage || (cursor?.isNotEmpty() ?: false)
 
-        val req = initialRequest.copy {
-            this.marker = cursor
-        }
+   override suspend operator fun next(): ListFunctionsResponse {
+      if (!hasNext()) throw NoSuchElementException("no pages remaining")
 
-        val result = client.listFunctions(req)
-        isFirstPage = false
-        cursor = result.nextMarker
-        return result
-    }
-    
-    // "functions" is the member targeted by the trait's `items` property
-    val functions: Paginator<List<FunctionConfiguration>?> = map(ListFunctionsResponse::functions)
+      val req = initialRequest.copy {
+         this.marker = cursor
+      }
+
+      val result = client.listFunctions(req)
+      isFirstPage = false
+      cursor = result.nextMarker
+      return result
+   }
+
+   // "functions" is the member targeted by the trait's `items` property
+   val functions: SdkAsyncIterable<FunctionConfiguration> = mapNotNull(ListFunctionsResponse::functions).flatten()
```

Comment:
If it were a sparse list then yes it would be `SdkAsyncIterable<T?>`. So no it wouldn't be every circumstance. I'm not necessarily convinced this is the right thing to do but I agree it is probably the most ergonomic to flatten and remove nulls.

Reply:
Maybe I'm missing something but can't we know definitively via the [presence of `sparse`](https://awslabs.github.io/smithy/1.0/spec/core/model.html#list)?

Reply:
Yes we can, I _think_ Ian was just asking if this would always be generated this way or if there are cases (like sparse) where we would generate something different. 

I'll try and clarify in the doc this call out.